util/src/openssl_utils.cpp

```diff
@@ -0,0 +1,266 @@
+  certStr.resize(certLen);
+  const auto res = BIO_read(outbio.get(), (void*)&(certStr.front()), certLen);
+  if (OPENSSL_FAILURE == res or OPENSSL_ERROR == res) {
+    LOG_ERROR(OPENSSL_LOG, "Failed to read data from the BIO certifiate object.");
+    return {};
+  }
+  return certStr;
+}
+
+bool CertificateUtils::verifyCertificate(X509& cert, const string& public_key) {
+  UniquePKEY pub_key(EVP_PKEY_new());
+  UniqueOpenSSLBIO pub_bio(BIO_new(BIO_s_mem()));
+
+  if (BIO_write(pub_bio.get(), static_cast<const char*>(public_key.c_str()), public_key.size()) <= 0) {
+    return false;
+  }
+  if (!PEM_read_bio_PUBKEY(pub_bio.get(), reinterpret_cast<EVP_PKEY**>(&pub_key), nullptr, nullptr)) {
+    return false;
+  }
+  return (bool)X509_verify(&cert, pub_key.get());
+}
+
+bool CertificateUtils::verifyCertificate(const X509& cert_to_verify,
+                                         const string& cert_root_directory,
+                                         uint32_t& remote_peer_id,
+                                         string& conn_type,
+                                         bool use_unified_certs) {
+  // First get the source ID
+  static constexpr size_t SIZE = 512;
+  string subject(SIZE, 0);
+  X509_NAME_oneline(X509_get_subject_name(&cert_to_verify), subject.data(), SIZE);
+
+  int peerIdPrefixLength = 3;
+  std::regex r("OU=\\d*", std::regex_constants::icase);
+  std::smatch sm;
+  regex_search(subject, sm, r);
+  if (sm.length() <= peerIdPrefixLength) {
+    LOG_ERROR(OPENSSL_LOG, "OU not found or empty: " << subject);
+    return false;
+  }
+
+  auto remPeer = sm.str().substr(peerIdPrefixLength, sm.str().length() - peerIdPrefixLength);
+  if (0 == remPeer.length()) {
+    LOG_ERROR(OPENSSL_LOG, "OU empty " << subject);
+    return false;
+  }
+
+  uint32_t remotePeerId;
+  try {
+    remotePeerId = stoul(remPeer, nullptr);
+  } catch (const std::invalid_argument& ia) {
+    LOG_ERROR(OPENSSL_LOG, "cannot convert OU, " << subject << ", " << ia.what());
+    return false;
+  } catch (const std::out_of_range& e) {
+    LOG_ERROR(OPENSSL_LOG, "cannot convert OU, " << subject << ", " << e.what());
+    return false;
+  }
+  remote_peer_id = remotePeerId;
+  string CN;
+  CN.resize(SIZE);
+  X509_NAME_get_text_by_NID(X509_get_subject_name(&cert_to_verify), NID_commonName, CN.data(), SIZE);
+  string cert_type = "server";
+  if (CN.find("cli") != string::npos) {
+    cert_type = "client";
+  }
+  conn_type = cert_type;
+
+  // Get the local stored certificate for this peer
+  string local_cert_path =
+      (use_unified_certs)
+          ? cert_root_directory + "/" + std::to_string(remotePeerId) + "/" + "node.cert"
+          : cert_root_directory + "/" + std::to_string(remotePeerId) + "/" + cert_type + "/" + cert_type + ".cert";
+  auto deleter = [](FILE* fp) {
+    if (nullptr != fp) {
+      fclose(fp);
+    }
+  };
+  std::unique_ptr<FILE, decltype(deleter)> fp(fopen(local_cert_path.c_str(), "r"), deleter);
```

Comment:
since this is a duplication, you should probably extract the c-file wrapper to some utility class.

Reply:
Changing to `std::unique_ptr<FILE, decltype(&fclose)> fp(tmpfile(), fclose);`